test-data/unit/check-classes.test

```diff
@@ -1367,6 +1367,54 @@ a = A()
 a.f = ''
 a.f = 1 # E: Incompatible types in assignment (expression has type "int", variable has type "str")
 
+[case testSettingDescriptorWithOverloadedDunderSet1]
+from typing import Any, overload, Union
+class D:
+    @overload
+    def __set__(self, inst: Any, value: str) -> None: pass
+    @overload
+    def __set__(self, inst: Any, value: int) -> None: pass
+    def __set__(self, inst: Any, value: Union[str, int]) -> None: pass
+class A:
+    f = D()
+a = A()
+a.f = ''
+a.f = 1
+a.f = 1.5
```

Comment:
Can you add a `# E` comment here as a hint to the reader about where the error happens, like this:

```
a.f = 1.5  # E
```

(This is only a comment and has no semantic significance.)